### ISSUE TITLE
default local lib?

### ISSUE THREAD

**Issue by steromano** [Mullefa/virtualenv#7]
The fact that `locallib` requires you to install all your non-base packages locally is good for production but can be a bit annoying sometimes for exploratory stuff. What if I just want to try out some package (maybe with a bunch of dependencies) but not necessarily inside a project?
What I'm currently doing is having "generic" local lib, which is basically the equivalent of my global library, and if I'm not working on a specific project I just activate that one.
I was wondering if it could make sense to have an environment variable pointing to your generic library, and have `use_local_lib` default to that if it doesn't find a local library in the working directory.
Or alternatively providing a `use_global_lib` function of some sort.
Thoughts?

**Comment by Mullefa** [comment]
Great idea as per. Simple (in a good way) solution to an annoying problem.

Are you think the file path for the global library could be set as an
environment variable in `.Renviron`?

If something like this was implemented I would also edit the
`create_.Rprofile()` function with an argument to specify whether the project
should be set up to use the local or global library.

**Comment by steromano** [comment]
Hadn't seen `create_.Rprofile`, quite handy.
An other option I'll mention is to move away from the idea that things will depend heavily on the working directory, and do something more like `virtualenvwrapper`: put all your local libraries in the a place that `locallib` knows about, and activate your local environment from anywhere by name (something like `use_local_lib("myproject")` or maybe `locallib:::use("myproject")`). At that point typing something like `use_local_lib("global")` seems a lot more reasonable than `use_local_lib("<a path which I have to remember>").`

**Comment by Mullefa** [comment]
Interesting, can see some of the merits for this. Do you think it would be too convoluted to provide both options? i.e. either a library in the root directory of the project, or a directory (`locallib/inst/libraries` which would flatten to `locallib/libraries` on installation) where named local libraries could be created and referenced?

**Comment by steromano** [comment]
But then how would `create_local_lib` work? One of the nice things of `virtualenvwrapper` is that you can both create and activate environments on the fly from anywhere just using their name, e.g.

```
mkvirtualenv mynewenv
workon mynewenv
```

And then you get some added bonuses like being able to list all available environments and things like that, because all environments are in the same place.

If you decide to go down this way, a possible solution for what you want could be as simple as having a `.Rprofile` in the root directory of the project that activates the desired environment.

**Comment by Mullefa** [comment]
Are you experiencing a pattern where you would like many projects per library, many times over? I think this would add weight to your suggestion. 

On a related note, how would `thaw()`ing a project which references a local library by name work? To be consistent with the `.Rprofile` of the project, a local library would with the same name would have to be created in `locallib/libraries/`. How would name conflicts be resolved?

**Comment by steromano** [comment]
Yeah I understand your point - the python philosophy is actually quite different from the `locallib` one, in that the `requirements.txt` file is not tied to a specific environment. If you adopt that approach, than `thaw()` would simply install everything in whatever local library is currently active , and it would be left to the user to create a new one with a suitable name beforehand if he so desires.

As for your first question - it kind of relates to the original issue: it is handy to have one or more reusable libraries (think `use_local_lib("datascience")`, `use_local_lib("pkgdev")`, `use_local_lib("hadley")`), often for exploratory purposes, that you can just switch on without having to install a bunch of stuff every time. I guess what I'm asking for is just a simpler way to access "generic" libraries.
Having just one (as I originally suggested) would already be pretty good. But thinking more about it, the possibility of supporting both mechanisms, as you suggested, probably deserves more thought.

**Comment by steromano** [comment]
Regarding the twofold functionality: my initial inclination would be to leave `locallib` as it currently is, but add on top of it a notion of "global" library, which works the `virtualenvwrapper` way:

```
create_global_lib("foo")
use_global_lib("foo")
```

basically by name instead of by path. 
I know that one of the initial ideas behind `locallib` was to simplify away the local/global distinction, but you could make this as an almost unrelated optional feature. It could almost be a different package, except for the fact that it should be impossible to activate a global lib when a local one is activated and vice-versa.
Then you could keep using `locallib` as you currently would to create isolated production-like environments, where libraries are "immutable", but you could still (optionally) access one or more "mutable" libraries for generic/exploratory bits of work, preserving the interactive aspect.

Let me know what you think, if you like the idea I might try and find the time to give it a shot.

**Comment by Mullefa** [comment]
:+1:
